src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+        KeysetData prev_data(fabric_index, prev_id);
+        ReturnErrorOnFailure(prev_data.Load(mStorage));
+        prev_data.next = keyset.next;
+        ReturnErrorOnFailure(prev_data.Save(mStorage));
+    }
+    else
+    {
+        // Remove first keyset
+        fabric.first_keyset = keyset.next;
+    }
+    if (fabric.keyset_count > 0)
+    {
+        fabric.keyset_count--;
+    }
+    // Update fabric info
+    return fabric.Save(mStorage);
+}
+
+GroupDataProvider::KeysetIterator * GroupDataProviderImpl::IterateKeysets(chip::FabricIndex fabric_index)
+{
+    VerifyOrReturnError(mInitialized, nullptr);
+    return mKeysetIterators.CreateObject(*this, fabric_index);
+}
+
+GroupDataProviderImpl::KeysetIteratorImpl::KeysetIteratorImpl(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index) :
+    mProvider(provider), mFabric(fabric_index)
+{
+    FabricData fabric(fabric_index);
+    if (CHIP_NO_ERROR == fabric.Load(provider.mStorage))
+    {
+        mNextId = fabric.first_keyset;
+        mCount  = fabric.keyset_count;
+        mIndex  = 0;
+    }
+}
+
+size_t GroupDataProviderImpl::KeysetIteratorImpl::Count()
+{
+    return mCount;
+}
+
+bool GroupDataProviderImpl::KeysetIteratorImpl::Next(Keyset & item)
+{
+    VerifyOrReturnError(mIndex < mCount, false);
+
+    KeysetData keyset(mFabric, mNextId);
+    VerifyOrReturnError(CHIP_NO_ERROR == keyset.Load(mProvider.mStorage), false);
+
+    mIndex++;
+    mNextId            = keyset.next;
+    item.keyset_id     = keyset.keyset_id;
+    item.policy        = keyset.policy;
+    item.num_keys_used = keyset.num_keys_used;
+    memcpy(item.epoch_keys, keyset.epoch_keys, sizeof(item.epoch_keys));
+    return true;
+}
+
+void GroupDataProviderImpl::KeysetIteratorImpl::Release()
+{
+    mProvider.mKeysetIterators.ReleaseObject(this);
+}
+
+//
+// Fabrics
+//
+
+CHIP_ERROR GroupDataProviderImpl::RemoveFabric(chip::FabricIndex fabric_index)
+{
+    FabricData fabric(fabric_index);
+    VerifyOrReturnError(CHIP_NO_ERROR == fabric.Load(mStorage), CHIP_ERROR_KEY_NOT_FOUND);
+
+    // Remove Group Mappings
+
+    EndpointData endpoint(fabric_index, fabric.first_endpoint);
+    size_t endpoint_count = 0;
+
+    do
+    {
+        if (CHIP_NO_ERROR != endpoint.Load(mStorage))
+        {
+            break;
+        }
+        RemoveAllGroupMappings(fabric_index, endpoint.id);
+        endpoint.id = endpoint.next;
+    } while ((++endpoint_count < fabric.endpoint_count) && (kInvalidEndpointId != endpoint.id));
+
+    // Remove States
+
+    StateListData states;
+
+    // Load state list info
+    states.Load(mStorage);
+
+    StateData state(states.first);
+    size_t state_index = 0;
+
+    while (state_index < states.count)
+    {
+        if (CHIP_NO_ERROR != state.Load(mStorage))
+        {
+            break;
+        }
+        if (state.fabric_index == fabric_index)
+        {
+            // Removing the state shifts down the upper states
```

Comment:
Is this comment still valid? I don't think up/down is valid any longer now that we have linked lists rather than flat lists.

Reply:
It is unfortunately valid. RemoveGroupState() is part of the public API, and uses actual array index as input, so if you remove element N, the N+1 element becomes N. In consequence, you need to be careful of not increasing the index when removing, otherwise you'll skip the entries next to the removed ones. I have rephrased this comment as:
```
// Remove the N state from the array. State N+1 becomes N
```